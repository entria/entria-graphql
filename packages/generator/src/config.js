/* @flow */
import path from 'path';
import merge from 'lodash.merge';
import fs from 'fs';
import pkgDir from 'pkg-dir';

const rootPath = pkgDir.sync('.') || '.';

let cacheConfig = null;

/**
 * Parse `.graphqlrc` config file and retrieve its contents
 * @param filePath {string} The path of the config file
 * @returns {*}
 */
const parseConfigFile = (filePath: string) => {
  const config = JSON.parse(fs.readFileSync(filePath, 'utf8'));

  const directories = Object.keys(config.directories).reduce((data, directory) => {
    if (directory === DIRECTORY_TYPE.SOURCE) {
      return {
        ...data,
        [directory]: `${rootPath}/${config.directories[directory]}`,
        [DIRECTORY_TYPE.SRC]: config.directories.source,
      };
    }

    return {
      ...data,
      [directory]: `${config.directories.source}/${config.directories.module}/${config.directories[directory]}`,
    };
  }, {});

  return {
    ...config,
    directories: {
      ...config.directories,
      ...directories,
    },
  };
};

export const DIRECTORY_TYPE = {
  SOURCE: 'source',
  SRC: 'src',
  MODULE: 'module',
  CONNECTION: 'connection',
  LOADER: 'loader',
  MODEL: 'model',
  MUTATION: 'mutation',
  MUTATION_TEST: 'mutation_test',
  TYPE: 'type',
  TYPE_TEST: 'type_test',
};

type DirectoryType = $Values<typeof DIRECTORY_TYPE>;
type Directories = {
  source: string,
  module: string,
  connection: string,
  loader: string,
  model: string,
  mutation: string,
  mutation_test: string,
  type: string,
  type_test: string,
  interface: string,
};

type Files = {
  schema: string,
};

type Config = {
  directories: Directories,
  files: Files,
}
/**
 * Get the `.graphqlrc` config file
 * @returns {object} The content of the config
 */
export const getCreateGraphQLConfig = (): Config => {
  // if (cacheConfig) return cacheConfig;

  // Use default config
  const defaultFilePath = path.resolve(`${__dirname}/graphqlrc.json`);

  const config = parseConfigFile(defaultFilePath);

  try {
    // Check if there is a `.graphqlrc` file in the root path
    const customConfig = parseConfigFile(`${rootPath}/.graphqlrc`);

    merge(config, customConfig);

    cacheConfig = config;

    // If it does, extend default config with it, so if the custom config has a missing line
    // it won't throw errors
    return config;
  } catch (err) {
    cacheConfig = config;
    // Return the default config if the custom doesn't exist
    return config;
  }
};

/**
 * Get a directory from the configuration file
 * @param directory {string} The name of the directory, e.g. 'source'/'mutation'
 * @returns {string} The directory path
 */
export const getConfigDir = (directory: DirectoryType) => getCreateGraphQLConfig().directories[directory];
