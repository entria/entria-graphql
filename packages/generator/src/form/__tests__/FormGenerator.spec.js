import helper from 'yeoman-test';
import assert from 'yeoman-assert';
import path from 'path';

import {
  getFileContent,
} from '../../../test/helpers';

import { getConfigDir } from '../../config';

const formGenerator = path.join(__dirname, '..');

it('generate a form file', async () => {
  const folder = await helper.run(formGenerator)
    .withArguments('Example')
    .toPromise();

  const destinationDir = getConfigDir('add');

  assert.file([
    `${destinationDir}/ExampleForm.js`,
  ]);

  expect(getFileContent(`${folder}/${destinationDir}/ExampleForm.js`)).toMatchSnapshot();
});
