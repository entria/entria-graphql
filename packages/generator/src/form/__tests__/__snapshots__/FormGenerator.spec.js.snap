// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate a form file 1`] = `
"// @flow
import * as React from 'react';
import { graphql, createFragmentContainer } from 'react-relay/compat';
import styled from 'styled-components';
import { withRouter } from 'react-router-dom';
import { withFormik } from 'formik';
import Yup from 'yup';

import type { ContextRouter } from 'react-router-dom';
import type { FormikProps } from 'formik';

import withSnackbar, { type SnackbarContextProps } from '../../hoc/withSnackbar';

import { routeTo } from '../../router/utils';
import { Button, ToggleFormik, InputFieldFormik } from '../common';

import type { ExampleForm_viewer } from './__generated__/ExampleForm_viewer.graphql';
import type { ExampleForm_node } from './__generated__/ExampleForm_node.graphql';

import ExampleAddMutation from './ExampleAddMutation';
import ExampleEditMutation from './ExampleEditMutation';

const FormContainer = styled.div\`
  display: flex;
  flex-wrap: wrap;
  padding-top: 25px;
  flex-direction: column;
\`;

const ButtonWrapper = styled.div\`
  display: flex;
  align-items: center;
  justify-content: flex-end;
\`;

const InputField = styled(InputFieldFormik).attrs({
  style: {
    marginTop: 25,
    marginBottom: 33,
    padding: 0,
    border: 'none',
    width: '100%',
    paddingBottom: 10,
    fontSize: 14,
  },
  floatingLabelStyle: {
    height: '16px',
    fontSize: '15px',
    lineHeight: '1.33',
    letterSpacing: 'normal',
    textAlign: 'left',
    color: '#607389',
  },
  underlineFocusStyle: {
    borderBottom: '2px solid rgb(72, 144, 229)',
  },
  underlineStyle: {
    borderBottom: '2px solid #D0D0D0',
  },
  underlineShow: true,
})\`\`;

const ToggleBox = styled.div\`
  display: flex;
  align-items: center;
\`;

const ToggleText = styled.p\`
  height: 16px;
  font-size: 12px;
  line-height: 1.33;
  letter-spacing: normal;
  text-align: left;
  color: #607389;
  width: 100px;
\`;

type OwnProps = {
  viewer: ExampleForm_viewer,
  node: ExampleForm_node,
};
type Props = OwnProps & FormikProps<OwnProps, Values> & SnackbarContextProps & ContextRouter;
class ExampleFormInnerFrom extends React.PureComponent<Props> {
  handleBack = () => {
    const { history, node } = this.props;

    if (node) return history.push(routeTo('example.view', { id: node.id }));

    return history.push(routeTo('example.list'));
  };

  render() {
    const { handleSubmit, isSubmitting, isValid } = this.props;

    return (
      <form onSubmit={handleSubmit}>
        <FormContainer>
          <InputField name=\\"name\\" floatingLabelText=\\"Nome\\" required={true} hintText={null} />
          <InputField
            name=\\"description\\"
            floatingLabelText=\\"Descrição\\"
            hintText={null}
            required={true}
          />
          <ToggleBox>
            <ToggleText>Está ativo?</ToggleText>
            <ToggleFormik name=\\"active\\" />
          </ToggleBox>
          <ButtonWrapper>
            <Button
              type=\\"button\\"
              label=\\"Cancelar\\"
              buttonType=\\"flat\\"
              labelStyle={{ color: '#757575' }}
              onClick={this.handleBack}
            />
            <Button
              type=\\"submit\\"
              label={isSubmitting ? 'Salvando..' : 'Salvar'}
              disabled={isSubmitting || !isValid}
              buttonType=\\"flat\\"
              labelStyle={{ color: '#4890E5' }}
              onClick={handleSubmit}
            />
          </ButtonWrapper>
        </FormContainer>
      </form>
    );
  }
}

type Values = {
  name: string,
  active: boolean,
};
const ExampleForm = withFormik({
  validationSchema: Yup.object().shape({
    name: Yup.string().required('É necessário um nome'),
    active: Yup.boolean(),
  }),
  mapPropsToValues: ({ node }: OwnProps) => {
    if (node) {
      return {
        name: node.name,
        active: node.active || false,
      };
    }

    return {
      name: '',
      active: true,
    };
  },
  handleSubmit: (values: Values, formikBag) => {
    const { setSubmitting, props } = formikBag;
    const { name, active } = values;
    const { history, showSnackbar, viewer, node, isSubmitting } = props;

    const isEditting = !!node;

    // TODO - handle Edit
    if (isSubmitting) return;

    if (isEditting) {
      const input = {
        id: node.id,
        name,
        active,
      };

      const onCompleted = ({ ExampleEdit }) => {
        setSubmitting(false);

        if (ExampleEdit.error) {
          showSnackbar({
            message: ExampleEdit.error,
          });

          return;
        }

        showSnackbar({
          message: 'Example editado com sucesso!',
        });

        return history.push(routeTo('example.view', { id: node.id }));
      };

      const onError = () => {
        setSubmitting(false);

        showSnackbar({
          message: 'Ocorreu um erro ao tentar editar o Produto.',
        });
      };

      ExampleEditMutation.commit(input, onCompleted, onError);

      return;
    }

    const input = {
      name,
      description,
      active,
    };

    const onCompleted = ({ ExampleAdd }) => {
      setSubmitting(false);

      if (ExampleAdd.error) {
        showSnackbar({
          message: ExampleAdd.error,
        });

        return;
      }

      showSnackbar({
        message: 'Example criado com sucesso!',
      });

      history.push(routeTo('example'));
    };

    const onError = () => {
      setSubmitting(false);

      showSnackbar({
        message: 'Ocorreu um erro ao tentar criar o Example.',
      });
    };

    ExampleAddMutation.commit(viewer.id, input, onCompleted, onError);
  },
})(ExampleFormInnerFrom);

const ExampleFormFormFragmentContainer = createFragmentContainer(ExampleForm, {
  node: graphql\`
    fragment ExampleForm_node on Node {
      id
      ... on Example {
        id
        name
        active
        # add more fields
      }
    }
  \`,
  viewer: graphql\`
    fragment ExampleForm_viewer on Viewer {
      id
    }
  \`,
});

export default withRouter(withSnackbar(ExampleFormFormFragmentContainer));
"
`;
